juju/juju
# feat(cmd-action): allow running actions without specifying unit number

Before this commit, `juju run` required the user to specify a unit by one of:
* full unit id, with number ("mysql/0")
* application name with leader alias ("mysql/leader")

After this commit, users can also specify the application name without any suffix and that will imply the action should run on the leader.  For example, `juju run mysql` resolves to `juju run mysql/leader`

A downside of this change is that, although the CLI still catches input errors, some cases are now ambiguous as to the *cause* of the error.  This is because application names, action names, and some key-value argument keys share the same schema.  For example, if a user does `juju run mysql/5 mysql action-name kvarg-missing-equal-sign`, we cannot definitively say whether "kvarg-missing-equal-sign" is a malformed key-value argument, or if "kvarg-missing-equal-sign" is the action name and instead "action-name" is actually a second application name.  In either case, users are alerted to there being *an* error.

## Checklist



  - usage instructions in the CLI are updated
  - the [relevant discourse page](https://juju.is/docs/juju/manage-actions#heading--run-an-action) was not changed because it does not go into this level of detail

## QA steps

on a Kubernetes controller
```
juju add-model kubeflow
juju deploy kubeflow-profiles --channel 1.8/stable --trust
juju run kubeflow-profiles create-profile username=x profilename=x
```

note that this action should execute on unit 0, but will fail with the message from the charm "Failed to apply PodDefaults"

## Documentation changes

CLI usage docs are updated

## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/2058262
